sled-agent/src/services.rs

```diff
@@ -2785,16 +2782,13 @@ impl ServiceManager {
                         SwitchService::MgDdm { mode } => {
                             info!(self.inner.log, "Setting up mg-ddm service");
                             smfh.setprop("config/mode", &mode)?;
-                            let info = self
-                                .inner
-                                .sled_info
-                                .get()
-                                .ok_or(Error::SledAgentNotReady)?;
-                            smfh.setprop("config/rack_uuid", info.rack_id)?;
-                            smfh.setprop(
-                                "config/sled_uuid",
-                                info.config.sled_id,
-                            )?;
+                            if let Some(info) = self.inner.sled_info.get() {
```

Comment:
Are we certain sled agent is going to come back around later and provide this info if a NotReady error is not returned here?

Reply:
Yeah, although every time I need to prove to myself it's true it takes forever to trace through how that happens. But this is the exact same pattern we already use for other services:

* MGS uses this to get the the `rack_id` once it's available
* wicketd uses this to get the rack subnet once it's available
* dendrite uses this to get the rack and sled IDs once they're available

Reply:
Tracking back how this happens:

1. We start up the bootstrap server. This does some initial ddmd config, and fairly early [spawns a bunch of tasks that run forever](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/bootstrap/pre_server.rs#L123-L136).
2. One of these tasks is the [hardware monitor](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/long_running_tasks.rs#L79-L81). Note that it returns a `sled_agent_started_tx` channel.
3. The main runloop of the hardware monitor task a `tokio::select!` that has an arm that [waits for a message on the `sled_agent_started` channel](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/hardware_monitor.rs#L135-L141). This calls `check_latest_hardware_snapshot()`, which [either activates or deactivates the switch based on whether the tofino driver is loaded](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/hardware_monitor.rs#L248-L252).
4. Calling `activate_switch` eventually lands us back in this function, where hopefully `sled_info` is set by the time we get here.

Separately, we can confirm that `sled_info` is set in this path:

1. `sled_info` is set by [ServiceManager::sled_agent_started](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/services.rs#L974-L985).
2. `ServiceManager::sled_agent_started` is called [from `SledAgent::new()`](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/sled_agent.rs#L551-L557).
3. `SledAgent::new()` is called [from `SledAgentServer::start()`](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/server.rs#L61-L70).
4. `SledAgentServer::start()` is called from [`start_sled_agent` in the bootstrap server](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/bootstrap/server.rs#L425-L435).
5. `start_sled_agent` is called from one of two places, depending on whether we're cold booting or waiting for RSS. In both cases, we send a handle to the sled agent on `sled_agent_started_tx` after doing so, which triggers the above process to "reactive" the switch (if we're a scrimlet). [cold boot case](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/bootstrap/server.rs#L238-L258),  [waiting for RSS case](https://github.com/oxidecomputer/omicron/blob/0ad5b9a38faf1511b06193fbc1d3e9f47ce17627/sled-agent/src/bootstrap/server.rs#L593-L614)

Reply:
I'm not sure if this path will actually work if it relies on `mgd` reacting to an SMF `refresh`.

Reply:
I'm 99% sure if the SMF manifest doesn't provide a `refresh` method, smf itself falls back to restarting the service. Let me see if I can find logs to confirm that from testing on madrid...

Reply:
Welp, I am incorrect:

```
[ Dec 29 12:29:24 Rereading configuration. ]
[ Dec 29 12:29:24 No 'refresh' method defined.  Treating as :true. ]
```

Should I change these `smfh.refresh()` calls to `smfh.restart()`? Or do we need to implement a real `refresh` on mgd/mg-ddm?

Reply:
Doing a restart would be undesirable. It would cause session flapping for both BGP and DDM. I'm working on a refresh capability in oxidecomputer/maghemite#184.

Reply:
My analysis above was incorrect, particularly this bit:

> 4\. Calling `activate_switch` eventually lands us back in this function, where hopefully `sled_info` is set by the time we get here.

If we land in `activate_switch` with an already-running switch zone, we do _not_ end up back in this function. It goes to `ensure_zone`, which has its own `match` to handle reconfiguring services ([starting here])(https://github.com/oxidecomputer/omicron/blob/f8c394f08191a95e1d83db33a8bf23a68b2c4a6b/sled-agent/src/services.rs#L3878). I'll add arms to that `match` for mgd and mg-ddm shortly.